daemon/src/order/maker.rs

```diff
@@ -0,0 +1,385 @@
+        let latest_offers = match self
+            .latest_offers
+            .send(xtra_libp2p_offer::maker::GetLatestOffers)
+            .await
+        {
+            Ok(latest_offers) => latest_offers,
+            Err(e) => {
+                bail!("Failed to retrieve latest offer from offers actor: {e:#}")
+            }
+        };
+
+        let offer = match latest_offers {
+            None => bail!("No latest offers available"),
+            Some(offers) => offers.pick_offer_to_take(offer_id),
+        };
+
+        match offer {
+            None => bail!("Offer with id {offer_id} not found in current offers"),
+            Some(offer) => Ok(offer),
+        }
+    }
+}
+
+#[xtra_productivity(message_impl = false)]
+impl Actor {
+    async fn handle(&mut self, msg: NewInboundSubstream, ctx: &mut xtra::Context<Self>) {
+        let NewInboundSubstream { peer, stream } = msg;
+
+        let mut framed = Framed::new(stream, JsonCodec::<MakerMessage, TakerMessage>::new());
+
+        let order = match self.await_order(&mut framed).await {
+            Ok(order) => order,
+            Err(e) => {
+                tracing::error!("Failed to receive order from taker: {e:#}");
+                return;
+            }
+        };
+
+        let (id, offer, quantity, leverage) = match order {
+            TakerMessage::PlaceOrder {
+                id,
+                offer: offer_id,
+                quantity,
+                leverage,
+            } => (id, offer_id, quantity, leverage),
+            TakerMessage::ContractSetupMsg(_) => {
+                tracing::error!("Unexpected message");
+                return;
+            }
+        };
+
+        tracing::info!(taker = %peer, %quantity, order_id = %id, "Taker wants to place an order");
+
+        // Reject the order if the offer cannot be found in the latest offers
+        let offer = match self.pick_offer(offer.id).await {
+            Ok(offer) => offer,
+            Err(e) => {
+                tracing::warn!("Rejecting taker order because unable to pick offer: {e:#}");
+
+                let future = async move {
+                    framed
+                        .send(MakerMessage::Decision(protocol::Decision::Reject))
+                        .await?;
+
+                    anyhow::Ok(())
+                };
+
+                tokio_extras::spawn_fallible(
+                    &ctx.address().expect("self to be alive"),
+                    future,
+                    move |e| async move {
+                        tracing::debug!(%peer, "Failed to send reject order message: {e}");
+                    },
+                );
+
+                return;
+            }
+        };
+
+        let oracle_event_id = offer.oracle_event_id;
+
+        let cfd = Cfd::from_order(
+            id,
+            &offer,
+            quantity,
+            Identity::new(x25519_dalek::PublicKey::from(
+                *b"hello world, oh what a beautiful",
+            )),
```

Comment:
oops, I was meant to reinstate this one! I only did it on the other side. we'll need to pass it in the message unfortunately, there's no other way. will add one more commit :)